java/cuvs-java/src/main/java22/com/nvidia/cuvs/spi/JDKProvider.java

```diff
@@ -32,12 +32,32 @@
 import java.lang.foreign.Arena;
 import java.lang.foreign.MemoryLayout;
 import java.lang.foreign.MemorySegment;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Objects;
 
 final class JDKProvider implements CuVSProvider {
 
+  private static final MethodHandle createNativeDataset$mh = createNativeDatasetBuilder();
+
+  static MethodHandle createNativeDatasetBuilder() {
+    try {
+      var lookup = MethodHandles.lookup();
+      var mt = MethodType.methodType(Dataset.class, MemorySegment.class, int.class, int.class);
+      return lookup.findStatic(JDKProvider.class, "createNativeDataset", mt);
+    } catch (NoSuchMethodException | IllegalAccessException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static Dataset createNativeDataset(
+      MemorySegment memorySegment, int size, int dimensions) {
+    return new DatasetImpl(null, memorySegment, size, dimensions);
+  }
+
```

Comment:
I have to say, this looks quite ingenious.

We'd like to be able to construct a `DataSet` from a `MemorySegment`.  But we'd like not to have to mention `MemorySegment` in the public API.

So instead, the method that constructs the `DataSet` is made private, and we return a `MethodHandle` to it.

To invoke the `MethodHandle`, we'd need an instance of the `MemorySegment`.  But by then, we're committed to Java 22.

Do I understand this correctly?

Reply:
That is correct @mythrocks